### DISCUSSION TITLE
Workaround for missing CorrelationId in HttpEndpoint in v3

### DISCUSSION THREAD

**Discussion by rolandgriesser** [elsa-workflows/elsa-core#4111] [Q&A] [answered]
According to the source code for v3-rc1, passing a `CorrelationId` in the `HttpEndpoint` won't have any effect due to
```
// TODO: Get correlation ID from query string or header.
var correlationId = default(string);
```

My current use case would create an external form that the user has to fill out, and I pass a callback url that points to a `HttpEndpoint` activity including a `CorrelationId` to that form. In v2, the workflow would stop and wait until the callback is called and then continue execution. With v3, this isn't possible right now due to the TODO mentioned above.
I've been looking into the `RunTask` activity, but I can't get a `TaskId` before entering the activity (I need to send it to the webform). Is there any other workaround I'm missing right now, or should I just opt for a custom activity in the meantime?

**Comment by sfmskywalker** [comment, marked as answer]
Hello @rolandgriesser,

Thanks for pointing out the `CorrelationId` issue. We've tackled this in Elsa 3 with the introduction of `IHttpCorrelationIdSelector`. This lets you extract correlation IDs from HTTP requests, bringing it in line with Elsa 2's behavior.

For the scenario where you're using an external form and a callback URL to an `HttpEndpoint` activity with a `CorrelationId`, the recent changes have got you covered.

Elsa 3 has two built-in selectors:
1. One pulls the correlation ID from the query string using the `correlationId` parameter.
2. The other gets it from the `X-Correlation-ID` header.

Need something custom? Implement the `IHttpCorrelationIdSelector` and register it with the `AddHttpCorrelationIdSelector` extension method.

Let us know if you have more questions or feedback.

Cheers
